### PR TITLE
[Feature][st-engine] Support for transform-v2 API

### PR DESCRIPTION
**Pull request by hailin0**
## Purpose of this pull request

Reference #2678

Release binary package:
    - Move engine starter jar from `lib/` to `starter/` dir
    - Outpout `seatunnel-transforms-v2.jar` to `lib/` dir
    
Maven modules:
    - Add seatunnel-transforms-v2 module
    - Add seatunnel-transforms-v2-e2e module
    
Engine Starter:
    - Add `lib/` to task execute classpath
    
ST-Engine:
    - Support for Split、Replace、Filter transform
    
E2E:
    - Copy `seatunnel-transforms-v2.jar` into container `lib/` dir
    - Improve get project root path for e2e test module

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by hailin0** [comment]
image

**Comment by hailin0** [comment]
@EricJoy2048 @ashulin @Hisoka-X @ic4y @CalvinKirs tx. PTAL

**Comment by hailin0** [comment]
@531651225  @TaoZex @TyrantLucifer @liugddx @laglangyue    th.  PTAL

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java [hunk 1]
```diff
@@ -52,7 +52,16 @@ public final class ContainerUtil {
...
+    public static final String PROJECT_ROOT_PATH = getProjectRootPath();
+
+    private static String getProjectRootPath() {
+        String e2eRootModuleDir = "seatunnel-e2e";
+        Path path = Paths.get(System.getProperty("user.dir"));
+        while (!path.endsWith(Paths.get(e2eRootModuleDir))) {
+            path = path.getParent();
+        }
+        return path.getParent().toString();
+    }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java:64]: Update get root path in e2e module

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/src/test/java/org/apache/seatunnel/e2e/transform/TestSuiteBase.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.testcontainers.containers.Network;
+
+@ExtendWith({
+    ContainerTestingExtension.class,
+    TestLoggerExtension.class,
+    TestCaseInvocationContextProvider.class
+})
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@DisabledOnContainer(value = {}, type = {EngineType.FLINK, EngineType.SPARK}, disabledReason = "TODO: Transform v2 translation to spark/flink isn't completed")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-transforms-v2-e2e/src/test/java/org/apache/seatunnel/e2e/transform/TestSuiteBase.java:40]: Now disable FLINK、SPARK

File: seatunnel-dist/src/main/assembly/assembly-bin.xml [hunk 3]
```diff
@@ -155,6 +155,19 @@
                 <include>org.apache.seatunnel:seatunnel-starter:jar</include>
             </includes>
             <outputFileNameMapping>${artifact.file.name}</outputFileNameMapping>
+            <outputDirectory>/starter</outputDirectory>
+            <scope>provided</scope>
+        </dependencySet>
+
+        <!-- ============ SeaTunnel Transforms Jars ============  -->
```
Comment (TaoZex) [inline seatunnel-dist/src/main/assembly/assembly-bin.xml]: How about use “SeaTunnel Transforms V2 Jars”？

**Comment by Hisoka-X** [comment]
@EricJoy2048 Are we should support repartition type transform in this version?

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/ReplaceTransform.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            return null;
+        }
+
+        if (isRegex) {
+            if (replaceFirst) {
+                return inputFieldValue.toString().replaceFirst(pattern, replacement);
+            }
+            return inputFieldValue.toString().replaceAll(pattern, replacement);
+        }
+        return inputFieldValue.toString().replace(pattern, replacement);
```
Comment (TaoZex) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/ReplaceTransform.java:105]: If there are multiple field names containing pattern, do we need to implement a method to replace the first pattern?

**Review by TaoZex — COMMENTED** [review]
Maybe we need to implement a subscript-based ReplaceTransform method that replaces the field name by the subscript.

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/CopyFieldTransform.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    protected SeaTunnelDataType getOutputFieldDataType() {
+        return srcFieldDataType;
+    }
+
+    @Override
+    protected Object getOutputFieldValue(SeaTunnelRowAccessor inputRow) {
+        return inputRow.getField(srcFieldIndex);
```
Comment (TaoZex) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/CopyFieldTransform.java]: Throw an exception in case srcField does not exist.

Comment (hailin0) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/CopyFieldTransform.java]: https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-api/src/main/java/org/apache/seatunnel/api/table/type/SeaTunnelRowType.java#L85

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/config/Common.java [hunk 6]
```diff
@@ -129,6 +130,30 @@ public static Path connectorDir() {
...
+
+    /**
+     * return lib jars, which located in 'lib/*'.
+     */
+    public static List<Path> getLibJars() {
+        Path libRootDir = Common.libDir();
+        if (!Files.exists(libRootDir) || !Files.isDirectory(libRootDir)) {
+            return Collections.emptyList();
+        }
+        return Arrays.stream(libRootDir.toFile().listFiles((dir, name) -> name.endsWith(".jar")))
```
Comment (Hisoka-X) [inline seatunnel-common/src/main/java/org/apache/seatunnel/common/config/Common.java]: The method seem like not support `lib/other/test.jar`? Maybe use `Files.walk` could be better.

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 7]
```diff
@@ -109,10 +108,22 @@ private void registerPlugin() {
                 }
             })
             .collect(Collectors.toList());
-
         pluginsJarDependencies.forEach(url -> FlinkCommon.ADD_URL_TO_CLASSLOADER.accept(Thread.currentThread().getContextClassLoader(), url));
 
+        List<URL> libJarDependencies = Common.getLibJars().stream()
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: Please reduce same code.

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/CopyFieldTransform.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
```
Comment (Hisoka-X) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/CopyFieldTransform.java]: Use deep copy？

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/ReplaceTransform.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    private boolean replaceFirst;
+
+    @Override
+    public String getPluginName() {
+        return "Replace";
+    }
+
+    @Override
+    protected void setConfig(Config pluginConfig) {
+        if (!pluginConfig.hasPath(KEY_REPLACE_FIELD)) {
```
Comment (Hisoka-X) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/ReplaceTransform.java]: Just use `CheckConfigUtil.checkAllExists`

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/SplitTransform.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    private String[] emptySplits;
+
+    @Override
+    public String getPluginName() {
+        return "Split";
+    }
+
+    @Override
+    protected void setConfig(Config pluginConfig) {
+        if (!pluginConfig.hasPath(KEY_SEPARATOR)) {
```
Comment (Hisoka-X) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/SplitTransform.java]: Just use `CheckConfigUtil.checkAllExists`

**Comment by EricJoy2048** [comment]
> @EricJoy2048 Are we should support repartition type transform in this version?

I suggest to add the support of repartition transform in the subsequent plans.

**Review by Hisoka-X — DISMISSED** [review]
LGTM

File: seatunnel-dist/pom.xml [hunk 11]
```diff
@@ -103,6 +103,13 @@
                     <version>${project.version}</version>
                     <scope>provided</scope>
                 </dependency>
+                <!-- transforms -->
+                <dependency>
+                    <groupId>org.apache.seatunnel</groupId>
+                    <artifactId>seatunnel-transforms-v2</artifactId>
+                    <version>${project.version}</version>
+                    <scope>provided</scope>
+                </dependency>
```
Comment (ashulin) [inline seatunnel-dist/pom.xml:112]: It should also need to be added to the `release` profile.
@CalvinKirs CC

**Review by ashulin — DISMISSED** [review]

File: seatunnel-dist/pom.xml (hunk 11 above)
Comment (hailin0) [reply seatunnel-dist/pom.xml:112]: fixed. added to `release` profile
@ashulin

**Review by EricJoy2048 — APPROVED** [review]
LGTM

**Review by Hisoka-X — APPROVED** [review]

**Review by TaoZex — APPROVED** [review]
+1

**Pull request merged**